cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -1750,6 +1764,26 @@ public void scanActiveRuns(ProgramId programId, Consumer<RunRecordDetail> consum
     }
   }
 
+  /**
+   * Scans active runs of the given program irrespective of the versions of the programs.
+   * Active runs means program run with status PENDING, STARTING, RUNNING, SUSPENDED or STOPPING.
+   * This method is similar to the {@link #scanActiveRuns(ProgramId, Consumer)},
+   * but takes a {@link ProgramReference} instead.
+   *
+   * @param programRef {@link ProgramReference} given program
+   * @param consumer a {@link Consumer} for processing the {@link RunRecordDetail} of each
+   *     active run.
+   */
+  public void scanActiveRunsAcrossVersions(ProgramReference programRef, Consumer<RunRecordDetail> consumer)
```

Comment:
I am wondering if this would potentially create long transactions. It's something we try to avoid, better to use io.cdap.cdap.internal.app.store.DefaultStore#scanActiveRuns

Reply:
Who is handling transactions here?

Reply:
On the other hand, we don't expect a lot of active runs... @albertshau WDYT?

Reply:
given you are interested if this number is higher than maxConcurrency, it makes sense to pass maxConcurrency+1 as a limit, so that we don't scan all if there is a lot. BTW: Given this check is done on every scheduled run, is it too costly? I think for direct run starting we are using a cache. @masoud-io I think you should know the cache machinery

Reply:
I think it should be ok since we don't expect a lot of active runs, and splitting it into multiple transactions could give an incorrect result. +1 to setting a limit.

Reply:
Added the limit / upperBound parameter to the scan method.